src/utils/schemaTree.js

```diff
@@ -0,0 +1,312 @@
+      const subschema = schema.properties[property];
+
+      subschema.name = property;
+
+      if (requiredList.has(property)) {
+        subschema.required = true;
+      }
+
+      createChildNode(rootNode, subschema, childIndex);
+    });
+  }
+}
+
+/**
+ * Fetch the reference schema defined by the refString
+ * @param {*} refString: value of $ref field within a schema
+ */
+function fetchRefSchema(schemaTree, refString) {
+  const [source, definitionPath] = refString.split('#');
+
+  /**
+   * Find the source for the reference
+   * TODO: add feature to fetch $refs in separate files or urls
+   */
+  if (source.length > 0) {
+  }
+
+  /** Find the definition within the source */
+  const parameters = definitionPath.split('/');
+  let ptr = schemaTree.schema;
+
+  parameters.forEach(parameter => {
+    if (parameter.length > 0) {
+      ptr = ptr[parameter];
+    }
+  });
+
+  return ptr;
+}
+
+/**
+ * Update the refNode's state to expanded form.
+ * @param {*} refDefaultNode: a refNode's defaultNode field
+ */
+export function expandRefNode(schemaTree, refDefaultNode) {
+  /**
+   * Create a clone of the schemaTree to maintain immutability.
+   */
+  const cloneTree = clone(schemaTree);
+
+  /**
+   * Traverse the clone tree using the refDefaultNode's path
+   * to find the corresponding ref node within the clone tree.
+   * ('nodePtr' ultimately points to a refNode with a structure
+   *   of defaultNode, expandedNode, isExpanded fields)
+   */
+  let nodePtr = cloneTree;
+
+  refDefaultNode.path.forEach(childIndex => {
+    nodePtr = nodePtr.children[childIndex];
+  });
+
+  /**
+   * Update the 'isExpanded' state of the ref node so that
+   * the ref row will now display an expanded version instead.
+   */
+  nodePtr.isExpanded = true;
+
+  /**
+   * If ref tree node has never referenced the $ref schema before,
+   * dynamically fetch the schema and store in expandedNode field
+   * in order to cache the referenced schema within the ref node.
+   */
+  if (!nodePtr.expandedNode) {
+    const refString = refDefaultNode.schema.$ref;
+    const expandedRefSchema = fetchRefSchema(schemaTree, refString);
+
+    nodePtr.expandedNode = createSchemaTree(expandedRefSchema, refDefaultNode.path);
+  }
+
+  return cloneTree;
+}
+
+/**
+ * Update the refNode's state to collapsed form.
+ * @param {*} refDefaultNode: a refNode's defaultNode field
+ */
+export function shrinkRefNode(schemaTree, refDefaultNode) {
+  /**
+   * Create a clone of the schemaTree to maintain immutability.
+   */
+  const cloneTree = clone(schemaTree);
+
+  /**
+   * Traverse the clone tree using the refDefaultNode's path
+   * to find the corresponding ref node within the clone tree.
+   * ('nodePtr' ultimately points to a refNode with a structure
+   *   of defaultNode, expandedNode, isExpanded fields)
+   */
+  let nodePtr = cloneTree;
+
+  refDefaultNode.path.forEach(childIndex => {
+    nodePtr = nodePtr.children[childIndex];
+  });

```

Comment:
See https://github.com/taskcluster/material-ui-json-schema-viewer/pull/54/files#r369504149 for reason instead of using `ramda.assocpath`